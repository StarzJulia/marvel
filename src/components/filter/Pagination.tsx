import React, {useState, useEffect, useContext} from 'react';
import {getPagesCount, getCurrentPage} from '../../scripts/pagination';
import {FilterContext} from '../../context/filterContext';

export default function Pagination() {
    let {total, count, offset, limit, change} = useContext(FilterContext);
    let [pages, setPages] = useState<number[]>([]);
    let [current, setCurrent] = useState(0);
    let [last, setLast] = useState(0);

    const createPagination = () => {
        let newPages = [];
        const curr = getCurrentPage(offset, count, limit);
        const lst = getPagesCount(total, limit);

        if (total > 0) {
            let start = Math.max(2, curr - 2);
            const end = Math.min(curr + 3, lst - 1);

            newPages.push(1);

            while (start <= end) {
                newPages.push(start);
                start += 1;
            }
            if (lst > 1)
                newPages.push(lst);
        }
        
        setCurrent(curr);
        setLast(lst);
        setPages(newPages);
    }

    const sendPaginationRequest = (page: number) => {
        change({
            offset: (page - 1) * limit
        });
    }

    useEffect(() => {
        createPagination();
    }, [offset, total, limit, count]);

    return (
        <div className="filter_pagination">
            {current > 1 &&
                <span
                    className="filter_pagination-page"
                    onClick={() => sendPaginationRequest(current - 1)}
                >
                    Prev
                </span>
            }
            {pages.map((page, index, pagesArr) => 
                <span key={index}> 
                    {index > 0 && (page - pagesArr[index - 1]) > 1 &&
                        <span>...</span>
                    }
                    <span 
                        className={(page == current ? "current" : "") + " filter_pagination-page"}
                        onClick={() => sendPaginationRequest(page)}
                    >
                        {page}
                    </span>
                </span>
            )}
            {current < last &&
                <span
                    className="filter_pagination-page"
                    onClick={() => sendPaginationRequest(current + 1)}
                >
                    Next
                </span>
            }
        </div>
    )
}