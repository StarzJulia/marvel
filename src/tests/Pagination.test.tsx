import React from 'react';
import { mount, configure } from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import Pagination from '../components/filter/Pagination';
import {getPagesCount, getCurrentPage} from '../scripts/pagination';

configure({ adapter: new Adapter() });

const initialProps = {
    total: 100,
    limit: 20,
    count: 20,
    offset: 40,
    paginate: jest.fn()
}

describe('Pagination', () => {
    const component = mount(<Pagination {...initialProps} />);

    it('should match the snapshot', () => {
        expect(component.html()).toMatchSnapshot();
    });

    it("number of pages should be 5", async () => {
        expect(getPagesCount(initialProps.total, initialProps.limit)).toBe(5);
    });

    it("current page should be 3", async () => {
        expect(getCurrentPage(initialProps.offset, initialProps.count, initialProps.limit)).toBe(3);
    });

    it('should render an pagination item', () => {
        expect(component.find('.filter_pagination-page').exists()).toBe(true);
    });

    it('on click, sendPaginationRequest function should be triggered on click event', () => {
        component.find('.filter_pagination-page').first().simulate('click');
        expect(initialProps.paginate).toHaveBeenCalled();
      });
});