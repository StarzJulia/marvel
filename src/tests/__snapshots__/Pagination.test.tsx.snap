// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination should match the snapshot 1`] = `"<div class=\\"filter_pagination\\"><span class=\\"filter_pagination-page\\">Prev</span><span><span class=\\" filter_pagination-page\\">1</span></span><span><span class=\\" filter_pagination-page\\">2</span></span><span><span class=\\"current filter_pagination-page\\">3</span></span><span><span class=\\" filter_pagination-page\\">4</span></span><span><span class=\\" filter_pagination-page\\">5</span></span><span class=\\"filter_pagination-page\\">Next</span></div>"`;
